source/retryable-writes/retryable-writes.rst

```diff
@@ -181,59 +181,94 @@ Retrying Write Commands
+is not possible and drivers MUST raise the original retryable error. In both
+cases, the caller is able to infer that an attempt was made.
+
+If the retry attempt also fails, drivers MUST update their topology according to
+the SDAM spec (see: `Error Handling`_) and raise any ensuing error to the user.
+In this case, the error from the retry attempt is just as relevant as the
+original retryable error and the caller is still able to infer that an attempt
+was made.
+
+Drivers MUST NOT attempt to retry a write command with the same transaction ID
+more than once.
+
+Consider the following pseudo-code:
+
 .. code:: typescript
 
   function executeRetryableWrite(command) {
+    /* Allow ServerSelectionException to propagate to our caller, which can then
+     * assume that no attempts were made. */
     server = selectServer("writable");
 
+    /* UnsupportedException will inform the caller of a configuration error in
+     * their application (i.e. using retryWrites=true with an old cluster) */
     if (server.getMaxWireVersion() < RETRYABLE_WIRE_VERSION) {
       throw new UnsupportedException();
     }
 
+    /* NetworkException and NotMasterException are both retryable errors. If
+     * caught, remember the exception, update SDAM accordingly, and proceed with
+     * retrying the operation. */
     try {
       return executeCommand(server, command);
-    } catch (NetworkException e) {
-      updateTopologyDescriptionForNetworkError(server, e);
-    } catch (NotMasterException e) {
-      updateTopologyDescriptionForNotMasterError(server, e);
+    } catch (NetworkException originalError) {
+      updateTopologyDescriptionForNetworkError(server, originalError);
+    } catch (NotMasterException originalError) {
+      updateTopologyDescriptionForNotMasterError(server, originalError);
     }
 
-    server = selectServer("writable");
+    /* If we cannot select a writable server, do not proceed with retrying and
+     * throw the original error. The caller can then infer that an attempt was
+     * made and failed. */
+    try {
+      server = selectServer("writable");
+    } catch (ServerSelectionException ignoredError) {
+      throw originalError;
+    }
 
-    // If the new server is too old, throw original error
+    /* If the server selected for retrying is too old, throw the original error.
+     * The caller can then infer that an attempt was made and failed. This case
+     * is very rare, and likely means that the cluster has mixed versions, just
+     * experienced a fail over, and we missed the error that retryable writes
+     * were not supported during the first attempt (3.6 feature compatibility
+     * could not have been enabled). */
     if (server.getMaxWireVersion() < RETRYABLE_WIRE_VERSION) {
-      throw e;
+      throw originalError;
     }
 
+    /* Allow any exception from the second attempt to propagate to our caller,
+     * as it is just as relevant (if not more) than the original error. */
```

Comment:
The spirit of SPEC-972 is all about ensuring the caller can infer whether an attempt was made or not (so they can anticipate the possibility that the write actually did commit and we just missed a response). That is why we save the original retryable error to re-throw if server selection or wire protocol checks fail for the second attempt.

In this last case, I think any possible error is on par with the original error. The caller can still infer an attempt was made, _and_ this error is more relevant than the original in that arose from the most recent topology state.

If anyone still feels that we should still only throw the original error here, please make your case. I'd be happy to revise it accordingly.